### PR TITLE
improvement: add http tags to request middleware

### PR DESCRIPTION
**Pull request by smoorpal**
==COMMIT_MSG==
Add http attributes to trace span tags in request middleware
==COMMIT_MSG==

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-server/265)

### CODE REVIEW

**Comment by smoorpal** [comment]
The naming of these variables can be found [here referenced by datadog](https://docs.datadoghq.com/logs/processing/attributes_naming_convention/#common-attributes). Columns like http.status_code and http.method are populated as top level params for datadog.

**Comment by smoorpal** [comment]
This has been tested with the expected results in staging.

File: witchcraft/internal/middleware/request_test.go [hunk 1]
```diff
@@ -102,6 +105,32 @@ func TestCombinedMiddleware(t *testing.T) {
...
+						"http.url_details.path": objmatcher.NewEqualsMatcher("/"),
+					},
+				),
+			}),
+		}
+		entries, err := logreader.EntriesFromContent(trcOutput.Bytes())
+		assert.NoError(t, err, "unexpected error when unmarshalling trace output")
+		assert.NoError(t, matcher.Matches(map[string]interface{}(entries[0])), "unexpected content in trace output")
+	}()
+
```
Comment (smoorpal) [inline witchcraft/internal/middleware/request_test.go]: Previously there was no test that checked the expected trace output. This checks for a single span with expected name and tags.

File: witchcraft/internal/tcpjson/tcp_logger_test.go [hunk 2]
```diff
@@ -103,7 +103,7 @@ func TestWriteFromSvc1log(t *testing.T) {
 	err = gotPayload.UnmarshalJSON(gotEnvelope.Payload)
 	require.NoError(t, err)
 	assert.Equal(t, "this is a test", gotPayload.Message)
-	assert.Equal(t, logging.LogLevelDebug, gotPayload.Level)
+	assert.Equal(t, logging.New_LogLevel(logging.LogLevel_DEBUG), gotPayload.Level)
```
Comment (smoorpal) [inline witchcraft/internal/tcpjson/tcp_logger_test.go]: WGS needs WGL to be upgraded to `1.10` in order to write out trace tags properly. There [appears to be a conjure generation change](https://github.com/palantir/witchcraft-go-logging/compare/v1.9.0...v1.10.0) included in 1.10. 

This has meant that the enum names and the structure of the log level are now different. I included these changes here but can do it in a precommit if you prefer.

Comment (smoorpal) [reply witchcraft/internal/tcpjson/tcp_logger_test.go]: In fact here is that precommit https://github.com/palantir/witchcraft-go-server/pull/266

**Review by bmoylan — CHANGES_REQUESTED** [review]
> Reviewed 4 of 7 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 4 of 7 files at r4, 1 of 1 files at r5, 3 of 4 files at r6, 2 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-server/265)* status: 14 of 15 files reviewed, 3 unresolved discussions (waiting on @smoorpal)

___
*[witchcraft/internal/middleware/request.go, line 137 at r7](https://reviewable.io/reviews/palantir/witchcraft-go-server/265#-MVvkwzV7ocsxeDpRxPI:-MVvkwzV7ocsxeDpRxPJ:bd7jbus) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/58feaacdfcca82b60842d2be78b092452c97ce3e/witchcraft/internal/middleware/request.go#L137)):*
> ```Go
> 			wtracing.WithSpanTag("http.url", req.URL.String()),
> 			wtracing.WithSpanTag("http.useragent", req.UserAgent()),
> 			wtracing.WithSpanTag("http.url_details.path", req.URL.Path),
> ```

The URL path is (1) unsafe and (2) varies by request. Can we use the path template here instead? That might require changing it to a RouteHandlerMiddleware and verifying that the ordering is still correct
___
*[witchcraft/internal/tcpjson/tcp_logger_test.go, line 106 at r6](https://reviewable.io/reviews/palantir/witchcraft-go-server/265#-MVvZwWTrL9qEUBfXTTM-r6-106:-MVvlUM2D9NCi2d8EG73:b-juwjen) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/e3a7bb0af7e3ff8bf8542acdf6f9287935b9f71b/witchcraft/internal/tcpjson/tcp_logger_test.go#L106)):*
<details><summary><i>Previously, smoorpal wrote…</i></summary><blockquote>

In fact here is that precommit https://github.com/palantir/witchcraft-go-server/pull/266
</blockquote></details>

Merged #266, let's rebase this branch on top

**Review by bmoylan — COMMENTED** [review]
> Reviewed 1 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-server/265)* status: all files reviewed, 3 unresolved discussions (waiting on @smoorpal)

**Review by bmoylan — COMMENTED** [review]
> Reviewed 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-server/265)* status: all files reviewed, 3 unresolved discussions (waiting on @smoorpal)

**Review by smoorpal — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-server/265)* status: all files reviewed, 3 unresolved discussions (waiting on @bmoylan and @smoorpal)

___
*[witchcraft/internal/middleware/request.go, line 137 at r7](https://reviewable.io/reviews/palantir/witchcraft-go-server/265#-MVvkwzV7ocsxeDpRxPI:-MVvmCM40vB7bNsTuAyA:b-pk968r) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/58feaacdfcca82b60842d2be78b092452c97ce3e/witchcraft/internal/middleware/request.go#L137)):*
<details><summary><i>Previously, bmoylan (Brad Moylan) wrote…</i></summary><blockquote>

The URL path is (1) unsafe and (2) varies by request. Can we use the path template here instead? That might require changing it to a RouteHandlerMiddleware and verifying that the ordering is still correct
</blockquote></details>

I actually don't have a hard requirement to add anything besides Method and Status. I was actually just trying to add everything that I thought was safe. I think the URL is useful in determining what endpoint was hit but I don't think worth the investment at this time if it requires a significant rework of the router
___
*[witchcraft/internal/tcpjson/tcp_logger_test.go, line 106 at r6](https://reviewable.io/reviews/palantir/witchcraft-go-server/265#-MVvZwWTrL9qEUBfXTTM-r6-106:-MVvmVXpErssh45Gh6yS:bramaim) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/e3a7bb0af7e3ff8bf8542acdf6f9287935b9f71b/witchcraft/internal/tcpjson/tcp_logger_test.go#L106)):*
<details><summary><i>Previously, bmoylan (Brad Moylan) wrote…</i></summary><blockquote>

Merged #266, let's rebase this branch on top
</blockquote></details>

I pulled #266 into this. Is that good or is there a different pr

**Comment by smoorpal** [comment]
___
*[witchcraft/internal/middleware/request.go, line 137 at r7](https://reviewable.io/reviews/palantir/witchcraft-go-server/265#-MVvkwzV7ocsxeDpRxPI:-MVvnL5BCMU9D_XH8-yB:b2fv6ze) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/58feaacdfcca82b60842d2be78b092452c97ce3e/witchcraft/internal/middleware/request.go#L137)):*
<details><summary><i>Previously, smoorpal wrote…</i></summary><blockquote>

I actually don't have a hard requirement to add anything besides Method and Status. I was actually just trying to add everything that I thought was safe. I think the URL is useful in determining what endpoint was hit but I don't think worth the investment at this time if it requires a significant rework of the router
</blockquote></details>

Updated just to be method, status, and useragent.

File: witchcraft/internal/middleware/request.go [hunk 3]
```diff
@@ -128,7 +129,11 @@ func NewRequestExtractIDs(
 
 		// retrieve existing trace info from request and create a span
 		reqSpanContext := b3.SpanExtractor(req)()
-		span := tracer.StartSpan("witchcraft-go-server request middleware", wtracing.WithParentSpanContext(reqSpanContext))
+		span := tracer.StartSpan("witchcraft-go-server request middleware",
```
Comment (tabboud) [inline witchcraft/internal/middleware/request.go:132]: Is there any convention for the name of the span as well?

**Review by bmoylan — APPROVED** [review]
> Reviewed 1 of 1 files at r9, 1 of 1 files at r10.
> Dismissed @smoorpal from a discussion.
> *[Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-server/265)* status: all files reviewed, 2 unresolved discussions (waiting on @smoorpal and @tabboud)

___
*[witchcraft/internal/middleware/request.go, line 132 at r9](https://reviewable.io/reviews/palantir/witchcraft-go-server/265#-MVrYbY8_Li8z0BC8rY1-r9-132:-MVvnnXx0sy1wX1H4fiu:b-3e22uk) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/8d172df51f3f2480e00efadeeba6642e48b320b2/witchcraft/internal/middleware/request.go#L132)):*
<details><summary><i>Previously, tabboud (Tony Abboud) wrote…</i></summary><blockquote>

Is there any convention for the name of the span as well?
</blockquote></details>

I wish there were, but not really. Probably fine to leave as-is until we want to enforce some broader consistency like `{import}/{type}.{method}` or something
___
*[witchcraft/internal/middleware/request_test.go, line 124 at r5](https://reviewable.io/reviews/palantir/witchcraft-go-server/265#-MVvoAIsCWu1I3_wpEUa:-MVvoAIsCWu1I3_wpEUb:bpubhi2) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/d183b645a5716604b5488081944fe25de733149b/witchcraft/internal/middleware/request_test.go#L124)):*
> ```Go
> 						"http.url":              objmatcher.NewEqualsMatcher("/"),
> 						"http.useragent":        objmatcher.NewRegExpMatcher(".*"),
> 						"http.url_details.path": objmatcher.NewEqualsMatcher("/"),
> ```

needs ./godelw format
___
*[witchcraft/internal/tcpjson/tcp_logger_test.go, line 106 at r6](https://reviewable.io/reviews/palantir/witchcraft-go-server/265#-MVvZwWTrL9qEUBfXTTM-r6-106:-MVvo3AUD7w7Z_ZDFuVn:bp1giu8) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/e3a7bb0af7e3ff8bf8542acdf6f9287935b9f71b/witchcraft/internal/tcpjson/tcp_logger_test.go#L106)):*
<details><summary><i>Previously, smoorpal wrote…</i></summary><blockquote>

I pulled #266 into this. Is that good or is there a different pr
</blockquote></details>

all good!

**Comment by smoorpal** [comment]
___
*[witchcraft/internal/middleware/request.go, line 132 at r9](https://reviewable.io/reviews/palantir/witchcraft-go-server/265#-MVrYbY8_Li8z0BC8rY1-r9-132:-MVvoA8MBEt_RpBjO00t:b-8arvj3) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/8d172df51f3f2480e00efadeeba6642e48b320b2/witchcraft/internal/middleware/request.go#L132)):*
<details><summary><i>Previously, bmoylan (Brad Moylan) wrote…</i></summary><blockquote>

I wish there were, but not really. Probably fine to leave as-is until we want to enforce some broader consistency like `{import}/{type}.{method}` or something
</blockquote></details>

AFAIK we have no conventions around span names right now. I think keeping the span name the same is probably best so anyone who is used to looking for this span doesn't lose it.  We can update sls-spec with guidance though.

**Comment by smoorpal** [comment]
___
*[witchcraft/internal/middleware/request_test.go, line 124 at r5](https://reviewable.io/reviews/palantir/witchcraft-go-server/265#-MVvoAIsCWu1I3_wpEUa:-MVvoT5G8qWqEfsRbHFD:b656z5s) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/d183b645a5716604b5488081944fe25de733149b/witchcraft/internal/middleware/request_test.go#L124)):*
<details><summary><i>Previously, bmoylan (Brad Moylan) wrote…</i></summary><blockquote>

needs ./godelw format
</blockquote></details>

my B. Thanks.

**Pull request merged**